File: Queue/priorityQueue.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct {
    int *array;
    int size;
    int capacity;
} PriorityQueue;

PriorityQueue* createPriorityQueue(int capacity) {
    PriorityQueue *pq = (PriorityQueue*)malloc(sizeof(PriorityQueue));
    pq->capacity = capacity;
    pq->size = 0;
    pq->array = (int*)malloc(pq->capacity * sizeof(int));
    return pq;
}

bool isFull(PriorityQueue* pq) {
    return (pq->size == pq->capacity);
}

bool isEmpty(PriorityQueue* pq) {
    return (pq->size == 0);
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

void heapifyUp(PriorityQueue* pq, int index) {
    while (index > 0 && pq->array[(index - 1) / 2] < pq->array[index]) {
        swap(&pq->array[(index - 1) / 2], &pq->array[index]);
        index = (index - 1) / 2;
    }
}

void heapifyDown(PriorityQueue* pq, int index) {
    int largest = index;
    int left = 2 * index + 1;
    int right = 2 * index + 2;

    if (left < pq->size && pq->array[left] > pq->array[largest]) {
        largest = left;
    }

    if (right < pq->size && pq->array[right] > pq->array[largest]) {
        largest = right;
    }

    if (largest != index) {
        swap(&pq->array[index], &pq->array[largest]);
        heapifyDown(pq, largest);
    }
}

void enqueue(PriorityQueue* pq, int item) {
    if (isFull(pq)) {
        printf("PriorityQueue is full\n");
        return;
    }
    pq->array[pq->size++] = item;
    heapifyUp(pq, pq->size - 1);
    printf("%d enqueued to priority queue\n", item);
}

int dequeue(PriorityQueue* pq) {
    if (isEmpty(pq)) {
        printf("PriorityQueue is empty\n");
        return -1;
    }
    int item = pq->array[0];
    pq->array[0] = pq->array[--pq->size];
    heapifyDown(pq, 0);
    return item;
}

int peek(PriorityQueue* pq) {
    if (isEmpty(pq)) {
        return -1;
    }
    return pq->array[0];
}

int main() {
    PriorityQueue* pq = createPriorityQueue(5);

    enqueue(pq, 3);
    enqueue(pq, 5);
    enqueue(pq, 1);
    enqueue(pq, 4);
    enqueue(pq, 2);

    printf("Front element is %d\n", peek(pq));
    printf("%d dequeued from priority queue\n", dequeue(pq));
    printf("Front element is %d\n", peek(pq));

    return 0;
}

File: Queue/PriorityQueue.java
package Queue;

public class PriorityQueue {
    private int[] array;
    private int size;
    private int capacity;

    public PriorityQueue(int capacity) {
        this.capacity = capacity;
        this.array = new int[capacity];
        this.size = 0;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public void enqueue(int item) {
        if (isFull()) {
            throw new IllegalStateException("PriorityQueue is full");
        }
        array[size++] = item;
        heapifyUp();
    }

    public int dequeue() {
        if (isEmpty()) {
            throw new IllegalStateException("PriorityQueue is empty");
        }
        int item = array[0];
        array[0] = array[--size];
        heapifyDown();
        return item;
    }

    public int peek() {
        if (isEmpty()) {
            throw new IllegalStateException("PriorityQueue is empty");
        }
        return array[0];
    }

    public int size() {
        return size;
    }

    private void heapifyUp() {
        int index = size - 1;
        while (index > 0 && array[parent(index)] < array[index]) {
            swap(index, parent(index));
            index = parent(index);
        }
    }

    private void heapifyDown() {
        int index = 0;
        while (leftChild(index) < size) {
            int largerChild = leftChild(index);
            if (rightChild(index) < size && array[rightChild(index)] > array[largerChild]) {
                largerChild = rightChild(index);
            }
            if (array[index] > array[largerChild]) {
                break;
            }
            swap(index, largerChild);
            index = largerChild;
        }
    }

    private int parent(int index) {
        return (index - 1) / 2;
    }

    private int leftChild(int index) {
        return 2 * index + 1;
    }

    private int rightChild(int index) {
        return 2 * index + 2;
    }

    private void swap(int index1, int index2) {
        int temp = array[index1];
        array[index1] = array[index2];
        array[index2] = temp;
    }

    public static void main(String[] args) {
        PriorityQueue pq = new PriorityQueue(5);
        pq.enqueue(3);
        pq.enqueue(5);
        pq.enqueue(1);
        pq.enqueue(4);
        pq.enqueue(2);

        System.out.println("Front element is: " + pq.peek());
        System.out.println("PriorityQueue size is: " + pq.size());

        pq.dequeue();
        pq.dequeue();

        System.out.println("Front element is: " + pq.peek());
        System.out.println("PriorityQueue size is: " + pq.size());
    }
}
